emote/algos/genrl/nn.py

```diff
@@ -0,0 +1,57 @@
+import torch
+
+from torch import Tensor, nn
+
+from emote.nn.action_value_mlp import ActionValueMlp
+from emote.nn.gaussian_policy import GaussianMlpPolicy
+
+
+class GenRLPolicy(GaussianMlpPolicy):
+    def __init__(
+        self,
+        decoder: nn.Module,
+        num_obs: int,
+        num_actions: int,
+        hidden_dims: list[int],
+    ):
+        super().__init__(num_obs, num_actions, hidden_dims)
+        self.num_actions = num_actions
+        self.decoder = decoder
+
+    def forward(self, obs: Tensor, epsilon: Tensor = None):
+        obs = obs.to(torch.float)
+        if len(obs.shape) == 1:
+            obs = obs.unsqueeze(dim=0)
+
+        if epsilon is not None:
+            epsilon = epsilon[:, : self.num_actions]
```

Comment:
@tgolsson I've removed all squeeze and unsqueeze things. The only place that we need to unsqueeze is in FrozenBatchNorm. And, if I correctly understood you, the problem with the regular BatchNorm is fixed now, and we can remove FrozenBatchNorm completely soon (if not now). 
However, I cannot fix the if-statement for epsilon. I must reduce the size of the tensor (epsilon) when it is not None. Otherwise, I'll get error later in the code when epsilon is multiplied by the policy std output. Please note that epsilon is generated for the original action space and with GenRL the action space is reduced to the size of the latent space. One possible solution can be to configure the code such that it generates epsilon with the right size. But, I do not know how. Any idea?

Reply:
Makes sense! A comment around that might be enough, but I'm not sure I understand the problem exactly. Is the distribution not per-action, but rather per-latent and the output action is derived directly from that?